### ISSUE TITLE
js output overwrites generated output by message name

### ISSUE THREAD

**Issue by shaunco** [bufbuild/buf#202]
js output isn't scoped by package names or directory structure, so if two proto files in different packages have a message with the same name in the same project, the generated js output overwrites any previous output:
* `test/first/sample1.proto` - with `package test.first` and a message named `something`
* `test/second/sample2.proto` - with `package test.second` and a message named `something`
* `buf.gen.yaml` - set to generate js

Run `buf generate -o gen` and only a single `gen/js/something.js` results. Using a file system monitor, it looks like both first and second were written, but second overwrites first with no warning or error from buf or protoc. `buf check lint` also didn't produce any warnings.

### Sample Project
[buf-js-test.zip](https://github.com/bufbuild/buf/files/5595107/buf-js-test.zip)

**Comment by bufdev** [comment]
Can you please post your code sample copy/pasted in the description? We’re not going to download a zip file.

**Comment by shaunco** [comment]
Not sure how this is better than the zip I attached, but here it is:

**buf.yaml**
```yaml
version: v1beta1
build:
  roots:
    - .
lint:
  use:
    - BASIC
    - FILE_LOWER_SNAKE_CASE
  except:
    - ENUM_NO_ALLOW_ALIAS
    - IMPORT_NO_PUBLIC
    - PACKAGE_AFFINITY
    - PACKAGE_DIRECTORY_MATCH
    - PACKAGE_SAME_DIRECTORY
breaking:
  use:
    - WIRE_JSON
```

**buf.gen.yaml**
```yaml
version: v1beta1
plugins:
  - name: js
    out: js
```

**test/first/sample1.proto**
```proto
syntax = "proto3";

package test.first;

option java_multiple_files = true;
option java_package = "org.test.first";
option csharp_namespace = "Test.First";
option go_package = "test.com/first;first";

message Something {
    string test_first = 1;
}
```

**test/second/sample2.proto**
```proto
syntax = "proto3";

package test.second;

option java_multiple_files = true;
option java_package = "org.test.second";
option csharp_namespace = "Test.Second";
option go_package = "test.com/second;second";

message Something {
    string test_second = 1;
}
```

**Comment by bufdev** [comment]
This isn't actually a bug, this is working as expected *given* the behavior of the `js` builtin plugin.  `buf generate` makes a parallel generation call per directory, roughly the equivalent of:

```bash
for dir in $(find "${ROOT}" -name '*.proto' -print0 | xargs -0 -n1 dirname | sort | uniq); do
  protoc -I "$ROOT $(find "${dir}" -name '*.proto')
done
```

With no options, the `js` bultin plugin generates the same filename for every directory, which is obviously not optimal. Luckily, this is easily solved by adding the `commonjs` option:

```
version: v1beta1
plugins:
  - name: js
    out: js
    opt: import_style=commonjs
```

This will result in different files being generated (which is how `protoc` plugins should generally operate).

**Comment by shaunco** [comment]
The protoc invocation per directory is also causing issues with [python-betterproro](https://github.com/danielgtaylor/python-betterproto/issues/168#issuecomment-737182807). While I've informed the betterproto authors, it would be great to get a command-line option or a per-plugin option in the `buf.gen.yaml` to do a single invocation of protoc for **all** protoc files instead of one invocation per directory. Perhaps something like --no-parallel-generation or 
```yaml
  - name: python_betterproto
    out: packages/python
    parallel: false
```

**Comment by bufdev** [comment]
We'll look into this, but note that we do not invoke once per source file, we invoke once per directory, which should work given the example you posted in that plugin.

**Comment by shaunco** [comment]
>but note that we do not invoke once per source file, we invoke once per directory

Sorry, that's what I get for writing up issues at 4am. I've corrected my comment on the betterproto side.

**Comment by bufdev** [comment]
See https://github.com/bufbuild/buf/issues/208

**Comment by bufdev** [comment]
This is released in v0.33.0. See https://docs.buf.build/generate-usage for details. Let us know if this works!
